### ISSUE TITLE
The position to add attention modules

### ISSUE THREAD

**Issue by x2ss** [kobiso/CBAM-tensorflow-slim#1]
Dear kobiso:

Thanks for your re-implementations of cbam-tensorflow-slim and SENet-tensorflow-slim. 

However, I find that the position you add cbam block is different from the paper "CBAM: Convolutional Block Attention Module" for the inception-resnet-v2. The same situation also happens for the  inception-resnet-v2 in SENet-tensorflow-slim. 

From my point of view, in both papers, i.e., "CBAM: Convolutional Block Attention Module" and "Squeeze-and-Excitation Networks", the attention modules should be placed between the original blocks of the inception-resnet-v2. So for the inception-resnet-v2, code of adding cbam and SENet modules would be as: 

net += scaled_up
 if activation_fn:
      net = activation_fn(net)
      if cbam_block:
          net=cbam_block(net, 'cbam_block') 
      if se_block:
          net = se_block(net, 'se_block')
which means that the attention modules would be added after "net += scaled_up".

Also in the inception-resnet-v2 of cbam-tensorflow-slim, you add the attention module in the  scaled_up branch as " scaled_up= cbam_block(scaled_up, 'cbam_block')" before  "net += scaled_up", while in the the inception-resnet-v2 of SENet-tensorflow-slim, you add the attention module in the  identity  branch as " net = se_block(net, 'se_block')" before  "net += scaled_up". 

I don't know if you make these modifications because these modifications show better performance in your own re-implementations. Or, do I make mistakes in understanding the way of adding these two attention modules to the inception-resnet-v2?

Thank you again for your hard work.

All the best.

**Comment by x2ss** [comment]
Since the  inception-resnet-v2 is a residual network, the right way to add  attention modules would be as scaled_up= cbam_block(scaled_up, 'cbam_block')" before "net += scaled_up" according to the paper "Squeeze-and-Excitation Networks". I have made a mistake in understanding the way of adding attention module.

**Comment by kobiso** [comment]
Actually, this is the thing made me confused as well.
Since *inception-resnet-v2* model has inception module which has residual summation and SENet paper did not specify where to put attention module, I was confused where to put it for best performance.
So I did experiments for both cases: 1) attention module after inception, 2) attention module before residual summation.
In the case of 2), the model became hard to train (I had to use small learning rate) and got poor performance.
And the case of 1) showed better performance than the other case.

One more to mention, I haven't tried to put attention module after reduction module in inception models.
I think it is good attempt to put attention module after reduction module and see if it gives any performance increase.

**Comment by x2ss** [comment]
I have the same problem when using inception with attention module (have to use small learning rate). 
Also, put attention module after reduction module in inception models is a inspiring idea which is worth trying. Thanks for sharing.

All the best.
